File: android/measure/src/test/java/sh/measure/android/fakes/FakeConfigProvider.kt
package sh.measure.android.fakes

import android.hardware.SensorManager
import sh.measure.android.config.ConfigProvider
import sh.measure.android.config.DynamicConfig
import sh.measure.android.config.MsrRequestHeadersProvider
import sh.measure.android.config.ScreenshotMaskLevel

internal class FakeConfigProvider : ConfigProvider {
    override val enableLogging: Boolean = false
    override var screenshotMaskHexColor: String = "#222222"
    override var screenshotCompressionQuality: Int = 25
    override var trackActivityIntentData: Boolean = false
    override var batchExportIntervalMs: Long = 3_000 // 3 seconds
    override var attachmentExportIntervalMs: Long = 500 // 500ms
    override var defaultHttpHeadersBlocklist: List<String> = emptyList()
    override var sessionBackgroundTimeoutThresholdMs: Long = 30_000 // 30 seconds
    override var maxEventNameLength: Int = 64
    override val customEventNameRegex: String = "^[a-zA-Z0-9_-]+$"
    override val maxUserDefinedAttributesPerEvent: Int = 100
    override var maxUserDefinedAttributeKeyLength: Int = 64
    override var maxUserDefinedAttributeValueLength: Int = 256
    override val autoStart: Boolean = true
    override val maxSpanNameLength: Int = 64
    override val maxCheckpointNameLength: Int = 64
    override val maxCheckpointsPerSpan: Int = 100
    override var maxInMemorySignalsQueueSize: Int = 30
    override val inMemorySignalsQueueFlushRateMs: Long = 3000
    override val maxAttachmentsInBugReport: Int = 5
    override val maxDescriptionLengthInBugReport: Int = 1000
    override val shakeAccelerationThreshold: Float = 2.5f * SensorManager.GRAVITY_EARTH
    override val shakeMinTimeIntervalMs: Long = 5000
    override val shakeSlop: Int = 2
    override val disallowedCustomHeaders: List<String> =
        listOf("Content-Type", "msr-req-id", "Authorization", "Content-Length")
    override val requestHeadersProvider: MsrRequestHeadersProvider? = null
    override var estimatedEventSizeInKb: Int = 2
    override var maxDiskUsageInMb: Int = 50
    override var enableFullCollectionMode: Boolean = true
    override var maxEventsInBatch: Int = 1000
    override var crashTimelineDurationSeconds: Int = 300
    override var anrTimelineDurationSeconds: Int = 300
    override var bugReportTimelineDurationSeconds: Int = 300
    override var traceSamplingRate: Float = 1f
    override var journeySamplingRate: Float = 1f
    override var screenshotMaskLevel: ScreenshotMaskLevel = ScreenshotMaskLevel.AllTextAndMedia
    override var cpuUsageInterval: Long = 3
    override var memoryUsageInterval: Long = 3
    override var crashTakeScreenshot: Boolean = true
    override var anrTakeScreenshot: Boolean = true
    override var launchSamplingRate: Float = 1f
    override var gestureClickTakeSnapshot: Boolean = true
    override var httpSamplingRate: Float = 1f
    override var httpDisableEventForUrls: List<String> = emptyList()
    override var httpTrackRequestForUrls: List<String> = emptyList()
    override var httpTrackResponseForUrls: List<String> = emptyList()
    override var httpBlockedHeaders: List<String> = emptyList()

    var shouldTrackHttpRequestBodyResult = true
    var shouldTrackHttpResponseBodyResult = true
    var shouldTrackHttpHeaderResult = true

    override fun shouldTrackHttpRequestBody(url: String): Boolean = shouldTrackHttpRequestBodyResult

    override fun shouldTrackHttpResponseBody(url: String): Boolean = shouldTrackHttpResponseBodyResult

    override fun shouldTrackHttpHeader(key: String): Boolean = shouldTrackHttpHeaderResult

    var shouldTrackHttpEventForUrl = true

    override fun shouldTrackHttpEvent(url: String): Boolean = shouldTrackHttpEventForUrl

    override fun setMeasureUrl(url: String) {
        // no-op
    }

    var dynamicConfig: DynamicConfig? = null
        private set

    override fun setDynamicConfig(config: DynamicConfig) {
        dynamicConfig = config
    }
}

File: flutter/packages/measure_flutter/test/utils/fake_config_provider.dart
import 'package:measure_flutter/src/config/config_provider.dart';
import 'package:measure_flutter/src/config/dynamic_config.dart';
import 'package:measure_flutter/src/config/screenshot_mask_level.dart';

class FakeConfigProvider implements ConfigProvider {
  List<String> _defaultHttpContentTypeAllowlist = [];
  List<String> _defaultHttpHeadersBlocklist = [];
  bool _enableLogging = false;
  int _maxCheckpointNameLength = 64;
  int _maxCheckpointsPerSpan = 100;
  int _maxSpanNameLength = 64;
  bool _autoStart = true;
  int _maxAttachmentsInBugReport = 5;
  int _maxDescriptionLengthInBugReport = 1000;
  int _screenshotCompressionQuality = 20;
  int _maxEventNameLength = 64;
  String _customEventNameRegex = '^[a-zA-Z0-9_-]+\$';
  int _maxUserDefinedAttributeValueLength = 100;
  int _maxUserDefinedAttributeKeyLength = 256;
  int _maxUserDefinedAttributesPerEvent = 256;
  double _traceSamplingRate = 100;
  bool _crashTakeScreenshot = true;
  bool _gestureClickTakeSnapshot = true;
  ScreenshotMaskLevel _screenshotMaskLevel = ScreenshotMaskLevel.none;
  final List<String> _httpDisableEventForUrls = [];
  final List<String> _httpTrackRequestForUrls = [];
  final List<String> _httpTrackResponseForUrls = [];
  List<String> _httpBlockedHeaders = [];
  bool shouldTrackHttpEventResult = true;
  Map<String, bool> shouldTrackHttpHeaderResults = {};
  bool shouldTrackHttpRequestBodyResult = true;
  bool shouldTrackHttpResponseBodyResult = true;

  // Getters
  @override
  List<String> get defaultHttpContentTypeAllowlist =>
      _defaultHttpContentTypeAllowlist;

  @override
  List<String> get defaultHttpHeadersBlocklist => _defaultHttpHeadersBlocklist;

  @override
  bool get enableLogging => _enableLogging;

  @override
  int get maxCheckpointNameLength => _maxCheckpointNameLength;

  @override
  int get maxCheckpointsPerSpan => _maxCheckpointsPerSpan;

  @override
  int get maxSpanNameLength => _maxSpanNameLength;

  @override
  int get maxAttachmentsInBugReport => _maxAttachmentsInBugReport;

  @override
  int get maxDescriptionLengthInBugReport => _maxDescriptionLengthInBugReport;

  @override
  int get screenshotCompressionQuality => _screenshotCompressionQuality;

  @override
  int get maxEventNameLength => _maxEventNameLength;

  @override
  String get customEventNameRegex => _customEventNameRegex;

  @override
  int get maxUserDefinedAttributeValueLength => _maxUserDefinedAttributeValueLength;

  @override
  int get maxUserDefinedAttributeKeyLength => _maxUserDefinedAttributeKeyLength;

  @override
  int get maxUserDefinedAttributesPerEvent => _maxUserDefinedAttributesPerEvent;

  @override
  bool get crashTakeScreenshot => _crashTakeScreenshot;

  @override
  bool get gestureClickTakeSnapshot => _gestureClickTakeSnapshot;

  @override
  List<String> get httpBlockedHeaders => _httpBlockedHeaders;

  @override
  List<String> get httpDisableEventForUrls => _httpDisableEventForUrls;

  @override
  List<String> get httpTrackRequestForUrls => _httpTrackRequestForUrls;

  @override
  List<String> get httpTrackResponseForUrls => _httpTrackResponseForUrls;

  @override
  ScreenshotMaskLevel get screenshotMaskLevel => _screenshotMaskLevel;

  @override
  double get traceSamplingRate => _traceSamplingRate;

  @override
  void setDynamicConfig(DynamicConfig dynamicConfig) {
    throw UnimplementedError();
  }

  @override
  void setMeasureUrl(String url) {
    throw UnimplementedError();
  }

  @override
  bool shouldTrackHttpEvent(String url) {
    return shouldTrackHttpEventResult;
  }

  @override
  bool shouldTrackHttpHeader(String key) {
    return shouldTrackHttpHeaderResults[key] ?? true;
  }

  @override
  bool shouldTrackHttpRequestBody(String url) {
    return shouldTrackHttpRequestBodyResult;
  }

  @override
  bool shouldTrackHttpResponseBody(String url) {
    return shouldTrackHttpResponseBodyResult;
  }

  // Setters
  set defaultHttpContentTypeAllowlist(List<String> value) =>
      _defaultHttpContentTypeAllowlist = value;

  set defaultHttpHeadersBlocklist(List<String> value) =>
      _defaultHttpHeadersBlocklist = value;

  set enableLogging(bool value) => _enableLogging = value;

  @override
  bool get autoStart => _autoStart;

  set maxCheckpointNameLength(int value) => _maxCheckpointNameLength = value;

  set maxCheckpointsPerSpan(int value) => _maxCheckpointsPerSpan = value;

  set maxSpanNameLength(int value) => _maxSpanNameLength = value;

  set autoStart(bool value) => _autoStart = value;

  set maxAttachmentsInBugReport(int value) =>
      _maxAttachmentsInBugReport = value;

  set maxDescriptionLengthInBugReport(int value) =>
      _maxDescriptionLengthInBugReport = value;

  set screenshotCompressionQuality(int value) =>
      _screenshotCompressionQuality = value;

  set customEventNameRegex(String value) => _customEventNameRegex = value;

  set maxEventNameLength(int value) => _maxEventNameLength = value;

  set maxUserDefinedAttributeValueLength(int value) => _maxUserDefinedAttributeValueLength = value;

  set maxUserDefinedAttributeKeyLength(int value) => _maxUserDefinedAttributeKeyLength = value;

  set maxUserDefinedAttributesPerEvent(int value) => _maxUserDefinedAttributesPerEvent = value;

  set crashTakeScreenshot(bool value) => _crashTakeScreenshot = value;

  set gestureClickTakeSnapshot(bool value) => _gestureClickTakeSnapshot = value;

  set traceSamplingRate(double value) => _traceSamplingRate = value;

  set screenshotMaskLevel(ScreenshotMaskLevel value) => _screenshotMaskLevel = value;

  set httpBlockedHeaders(List<String> value) => _httpBlockedHeaders = value;
}
